### PR TITLE
[quant][graphmode] Add support for functional linear

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40384 [quant][graphmode][refactor] InsertQuantDeQuantHelper
* #40383 [quant][graphmode] Add warning for debug option for add_scalar/mul_scalar
* **#40331 [quant][graphmode] Add support for functional linear**

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22162905](https://our.internmc.facebook.com/intern/diff/D22162905)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@z-a-f this should fix the problem you see when testing traced linear_relu

File: torch/csrc/jit/passes/quantization/insert_observers.cpp [hunk 1]
```diff
@@ -476,31 +476,34 @@ class InsertObserversHelper {
   // nn.Linear + nn.ReLU
   const PatternInfo nn_linear_nn_relu = PatternInfo::parse_from_str(
       R"(
-graph(%self, %linear, %relu, %input):
+graph(%input, %linear, %relu):
```
Comment (raghuramank100) [inline torch/csrc/jit/passes/quantization/insert_observers.cpp:479]: For my understanding, why was self part of the graph, when it is not used in the actual pattern?

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_observers.cpp:479]: yeah, that's why I removed it, it's not needed in the pattern.

**Review by vkuzo — APPROVED** [review]
lg

**Review by raghuramank100 — APPROVED** [review]

**Pull request closed without merging**